common/sgx/tlsverifier.c

```diff
@@ -172,3 +197,121 @@ oe_result_t oe_verify_attestation_certificate(
     oe_free(report);
     return result;
 }
+
+/**
+ * oe_verify_plugin_attestation_certificate
+ *
+ * This function perform a custom validation on the input certificate. This
+ * validation includes extracting an attestation evidence extension from the
+ * certificate before validating this evidence. An optional
+ * claim_verify_callback could be passed in for a calling client to further
+ * validate the claims of the enclave creating the certificate.
+ * @param[in] cert_in_der a pointer to buffer holding certificate contents
+ *  in DER format
+ * @param[in] cert_in_der_len size of certificate buffer above
+ * @param[in] claim_verify_callback callback routine for custom claim checking
+ * @param[in] arg an optional context pointer argument specified by the caller
+ * when setting callback
+ * @retval OE_OK on a successful validation
+ * @retval OE_VERIFY_FAILED on quote failure
+ * @retval OE_INVALID_PARAMETER At least one parameter is invalid
+ * @retval OE_FAILURE general failure
+ * @retval other appropriate error code
+ */
+oe_result_t oe_verify_plugin_attestation_certificate(
+    uint8_t* cert_in_der,
+    size_t cert_in_der_len,
+    oe_claim_verify_callback_t claim_verify_callback,
+    void* arg)
+{
+    oe_result_t result = OE_FAILURE;
+    oe_cert_t cert = {0};
+    uint8_t* report = NULL;
+    size_t report_size = 0;
+    oe_report_header_t* header = NULL;
+    uint8_t* pub_key_buf = NULL;
+    size_t pub_key_buf_size = KEY_BUFF_SIZE;
+    oe_claim_t* claims = NULL;
+    size_t claims_length = 0;
+
+    pub_key_buf = (uint8_t*)oe_malloc(KEY_BUFF_SIZE);
+    if (!pub_key_buf)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    result = oe_cert_read_der(&cert, cert_in_der, cert_in_der_len);
+    OE_CHECK_MSG(result, "cert_in_der_len=%d", cert_in_der_len);
+
+    // validate the certificate signature
+    result = oe_cert_verify(&cert, NULL, NULL, 0);
+    OE_CHECK_MSG(
+        result,
+        "oe_cert_verify failed with error = %s\n",
+        oe_result_str(result));
+
+    //------------------------------------------------------------------------
+    // Validate the report's trustworthiness
+    //------------------------------------------------------------------------
+
+    // determine the size of the extension
+    if (oe_cert_find_extension(
+            &cert, (const char*)oid_oe_report, NULL, &report_size) !=
```

Comment:
New oid needs to be defined for embedded evidence.

Reply:
I think it's fine to multiplex the oid for both quote-based report and evidence-based report. The buffer contains a oe_report_header_t that identifies which type it is. The added test case uses that to correctly verify a certificate based evidences.

Reply:
Thanks @CodeMonkeyLeet for the context. Added another oid X509_OID_FOR_EVIDENCE_EXT in report.h.

Reply:
I do not see the change in report.h?